SD_Card.h

```diff
@@ -0,0 +1,85 @@
+#include <SD.h>
+#include <SPI.h>
+
+#include "gps.h"
+#include "imu.h"
+
+#pragma once
+
+class SDCard {
+  private:
+    File m_sdCardDataFile;
+    static const int m_chipSelect = 4; // this is not the final value - SDCARD_SS_PIN is the correct one for the MKR Zero
+    static const String m_fileName = "rocketry_data";
+  public:
+    void writeToCSV( // this is equivalent 
+      const GPS::Coordinates& coords,
+      const IMU::vector3& accel,
+      const IMU::vector3& gyro,
+      float altitude,
+      int analogReading, // this is VOC
+      float humidity,
+      float temperature) {
+        // checks if SD card is open and good to be written to 
+        if (m_sdCardDataFile) {
+          String dataOutputString = "";
+
+          dataOutputString += String(coords.latitude);
+          dataOutputString += ",";
+          dataOutputString += String(coords.longitude);
+          dataOutputString += ",";
+          dataOutputString += String(coords.altitudeMSL);
+          dataOutputString += ",";
+          dataOutputString += String(coords.numSatellites);
+          dataOutputString += ",";
+          dataOutputString += String(coords.timestamp.hours) + " : " + String(coords.timestamp.minutes) + " : " + String(coords.timestamp.seconds) + " : " + String(coords.timestamp.milliseconds);
+          dataOutputString += ",";
+
+          dataOutputString += String(accel.x);
+          dataOutputString += ",";
+          dataOutputString += String(accel.y);
+          dataOutputString += ",";
+          dataOutputString += String(accel.z);
+          dataOutputString += ",";
+          
+          dataOutputString += String(gyro.x);
+          dataOutputString += ",";
+          dataOutputString += String(gyro.y);
+          dataOutputString += ",";
+          dataOutputString += String(gyro.z);
+          dataOutputString += ",";
+
+          dataOutputString += String(altitude);
+
+          dataOutputString += ",";
+
+          dataOutputString += String(analogReading);
+
+          dataOutputString += ",";
+
+          dataOutputString += String(humidity);
+
+          dataOutputString += ",";
+
+          dataOutputString += String(temperature);
+
+          m_sdCardDataFile.println(dataOutputString);
+        }
+      }
+
+    void sd_initialize() { // this is equivalent to setup()
+      if (!SD.begin(m_chipSelect)) {
+        Serial.println("SD card not present!!");
+
+        while (1);
```

Comment:
We probably don't want to lock up the whole board due to an SD error, do we? I see at least two major problems with this:

- If we don't insert the SD card until after this point is reached, it never checks again.
- With the SD card absent, we can still send data over the radio.

Notice what I do with the sensors:

- Each sensor has a `getStatus()` method that tells you whether it's ready.
- The loop checks `getStatus()` before using the sensor, and if the sensor isn't connected, it tries to initialize it again.

This way you don't accidentally attempt to write to a nonexistant SD card, but also if it isn't there it isn't the end of the world.

Reply:
@TheRoyalIrishman I don't feel like this was properly addressed. Removing the `SD.begin` call is a non-starter as now every read and write will fail -- the SPI communications were never set up. My suggestion is to look at how the altimeter, IMU, or humidity sensor's `initialize` and `getStatus` methods are implemented (the GPS is a bit different because it has a multi-phase initialization).